source/simulator/nullspace.cc

```diff
@@ -379,19 +393,37 @@ namespace aspect
     for (const auto &cell : dof_handler.active_cell_iterators())
       if (cell->is_locally_owned())
         {
-          fe.reinit (cell);
-          const std::vector<Point<dim> > &q_points = fe.get_quadrature_points();
+          if (limit_to_top_faces == false)
+            fe_values.reinit (cell);
+          else
+            {
+              // We only want the output at the top boundary, so only compute it if the current cell
+              // has a face at the top boundary.
+              bool cell_at_top_boundary = false;
+              for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+                if (cell->at_boundary(f) &&
+                    (geometry_model->translate_id_to_symbol_name (cell->face(f)->boundary_id()) == "top"))
+                  {
+                    cell_at_top_boundary = true;
```

Comment:
Assert(cell_at_top_boundary==false, ExcInternalError("error, more than one top surface found in a cell."));